sycl/doc/extensions/proposed/sycl_ext_oneapi_address_space_cast.asciidoc

```diff
@@ -0,0 +1,192 @@
+          typename ElementType>
+multi_ptr<ElementType, Space, DecorateAddress>
+static_address_space_cast(ElementType* ptr);
+
+template <access::address_space Space, access::decorated DecorateAddress,
+          typename ElementType>
+multi_ptr<ElementType, Space, DecorateAddress>
+dynamic_address_space_cast(ElementType* ptr);
+
+} // namespace sycl::ext::oneapi::experimental
+----
+
+[source,c++]
+----
+template <access::address_space Space, access::decorated DecorateAddress,
+          typename ElementType>
+multi_ptr<ElementType, Space, DecorateAddress>
+static_address_space_cast(ElementType* ptr);
+----
+_Preconditions_: `ptr` points to an object allocated in the address space
+designated by `Space`.
+
+_Returns_: A `multi_ptr` with the specified address space and decoration that
+points to the same object as `ptr`.
+
+[source,c++]
+----
+template <access::address_space Space, access::decorated DecorateAddress,
+          typename ElementType>
+multi_ptr<ElementType, Space, DecorateAddress>
+dynamic_address_space_cast(ElementType* ptr);
+----
+_Preconditions_: `ptr` is accessible to the calling work-item.
+
+_Returns_: A `multi_ptr` with the specified address space and decoration that
+points to the same object as `ptr` if `ptr` points to an object allocated in
+the address space designated by `Space`, and `nullptr` otherwise.
+
+_Diagnostics_: Implementations using address space deduction must issue a
+diagnostic if `ptr` does not point to an object allocated in the address space
+designated by `Space`.
+
+[NOTE]
+====
+The precondition prevents `dynamic_address_space_cast` from being used to
+reason about the address space of pointers originating from another work-item
+(in the case of `private` pointers) or another work-group (in the case of
+`local` pointers). Such pointers could not be dereferenced by the calling
+work-item, and it is thus unclear that being able to reason about the address
+space would be useful. Limiting the use of `dynamic_address_space_cast` to
+accessible pointers is expected to result in simpler and faster
+implementations.
+====
+
+
+== Implementation notes
+
+For SPIR-V backends, `static_address_space_cast` corresponds to
+`OpGenericCastToPtr`. `dynamic_address_space_cast` _may_ correspond to
+`OpGenericCastToPtrExplicit` -- there is currently some ambiguity regarding
+exactly how `OpGenericCastToPtrExplicit` is expected to behave, because the
+SPIR-V specification does not explain what it means for a cast to "fail".
+Since this extension is only experimental, we can likely implement
+`dynamic_address_space_cast` using `OpGenericCastToPtrExplicit` while we
+seek to clarify the SPIR-V specification.
+
+Generally speaking, it is expected that a `static_address_space_cast` can
+simply attach new decoration(s) to the raw pointer (or do nothing), while
+a `dynamic_address_space_cast` will have to inspect the address of the
+raw pointer to determine which region of memory it points to.
```

Comment:
FWIW, I thought through the implementation for a CPU target (e.g. the native CPU), and I think `dynamic_address_space_cast` can be implemented by keeping track of three pieces of information in thread-local storage:

* The base (highest address) of the calling thread's stack.
* The low bound of the calling work-item's SLM area.
* The high bound of the calling work-item's SLM area.

(The low bound of the calling thread's stack is easily available via the $RSP register.)

* A cast to `private_space` succeeds as long as the pointer is within the calling thread's stack.
* A cast to `local_space` succeeds as long as the pointer is within the calling work-item's SLM area.
* A cast to `global_space` succeeds as long as the pointer is **not** within either of the above two address ranges.

Reply:
I've added this in 3456c36ac2e89268eb5ff07b0232653ffc6fcfe9 so that it's easier for people to find, and added a note that some GPUs have instructions to perform these checks.